source/runtime_api.cpp

```diff
@@ -1019,62 +1019,122 @@ void reshade::runtime::set_preprocessor_definition_for_effect([[maybe_unused]] c
 	if (name == nullptr)
 		return;
 
-	const std::string effect_name_string = effect_name != nullptr ? effect_name : std::string();
+	if (effect_name == nullptr)
+		effect_name = "";
 
+	const int find_flag = *effect_name == '\0' ? 0b111 :
+		strncmp(effect_name, "PRESET", 6) == 0 ? 0b010 :
+		strncmp(effect_name, "GLOBAL", 6) == 0 ? 0b100 : 0b111;
+
+	int updated = 0;
 	if (value == nullptr || *value == '\0')
 	{
-		if (const auto it = std::remove_if(_global_preprocessor_definitions.begin(), _global_preprocessor_definitions.end(),
-				[name](const std::pair<std::string, std::string> &preset_definition) { return preset_definition.first == name; });
-			it != _global_preprocessor_definitions.end())
+		if (find_flag & 0b001)
+		{
+			if (const auto preset_it = _preset_preprocessor_definitions.find(effect_name); preset_it != _preset_preprocessor_definitions.end() && !preset_it->second.empty())
+				if (const auto it = std::remove_if(preset_it->second.begin(), preset_it->second.end(),
+					[name = std::string_view(name)](const std::pair<std::string, std::string> &definition) { return definition.first == name; }); it != preset_it->second.cend())
+					preset_it->second.erase(it, preset_it->second.end()), updated = 0x001;
+		}
+		if (find_flag & 0b010)
 		{
-			_global_preprocessor_definitions.erase(it, _global_preprocessor_definitions.end());
+			if (const auto preset_it = _preset_preprocessor_definitions.find({}); preset_it != _preset_preprocessor_definitions.end() && !preset_it->second.empty())
+				if (const auto it = std::remove_if(preset_it->second.begin(), preset_it->second.end(),
+					[name = std::string_view(name)](const std::pair<std::string, std::string> &definition) { return definition.first == name; }); it != preset_it->second.cend())
+					preset_it->second.erase(it, preset_it->second.end()), updated = 0x010;
 		}
+		if (find_flag & 0b100)
+		{
+			if (const auto it = std::remove_if(_global_preprocessor_definitions.begin(), _global_preprocessor_definitions.end(),
+				[name = std::string_view(name)](const std::pair<std::string, std::string> &definition) { return definition.first == name; });
+				it != _global_preprocessor_definitions.end())
+				_global_preprocessor_definitions.erase(it, _global_preprocessor_definitions.end()), updated = 0x100;
```

Comment:
This was probably supposed to be `updated = 0b100;`, rather than `update = 0x100;` (same for the two above). Maybe it would help to make this an flags enum, for readability?
Btw., I find code becomes easier to read with some more curly braces and semicolons 😉:
```c++
if (const auto it = std::remove_if(_global_preprocessor_definitions.begin(), _global_preprocessor_definitions.end(),
		[name = std::string_view(name)](const std::pair<std::string, std::string> &definition) { return definition.first == name; });
	it != _global_preprocessor_definitions.end())
{
	_global_preprocessor_definitions.erase(it, _global_preprocessor_definitions.end());
	updated = 0b100;
}
```